sql: panic: invalid datum type given: int, expected decimal

`select 0::int8 + 3989212821610048162::decimal // 8965304063540790833::int8`

```
panic: invalid datum type given: int, expected decimal [recovered]
        panic: panic while executing 1 statements: SELECT _::INT8 + (_::DECIMAL // _::INT8); caused by invalid datum type given: int, expected decimal

goroutine 410 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00083c000, 0x37e69a0, 0xc001416180, 0x2b51e40, 0xc0002b2660)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:721 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00083c000, 0x37e69a0, 0xc001416180)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:427 +0x61
panic(0x2b51e40, 0xc0002b2660)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:170 +0x25d
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc0007eea00, 0xc00083c460, 0x37e8c20, 0xc000c3ed70, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:190 +0x593
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x37e6a60, 0xc001535800, 0x37f1fe0, 0xc0007eea00, 0x37d8460, 0xc0014c8380)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:175 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc0007eea00, 0x37e6a60, 0xc001535800)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:801 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc00140c380, 0x37e6a60, 0xc001535800, 0x3198718, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:649 +0x209
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0007d5900, 0xc0005b1b60, 0xc000a6ad80, 0xc0016fa990, 0xc000457080, 0xc00083c520, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:252 +0x8a0
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0007d5900, 0x37e6a60, 0xc001535320, 0xc00083c520, 0xc0005b1b60, 0xc000a6ad80, 0x37e8c20, 0xc000c3ed70, 0xc000457080)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:795 +0x227
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00083c000, 0x37e6a60, 0xc001535320, 0xc00083c460, 0x3, 0x7fc6b3e2a290, 0xc000fc0630, 0x0, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1051 +0x26e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00083c000, 0x37e6a60, 0xc001535320, 0xc00083c460, 0x7fc6b3e2a290, 0xc000fc0630, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:900 +0x644
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00083c000, 0x37e6a60, 0xc001535320, 0x37ec560, 0xc0018b1ac0, 0xc000fa20e9, 0x4a, 0x0, 0x0, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:438 +0xe49
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00083c000, 0x37e6a60, 0xc001535320, 0x37ec560, 0xc0018b1ac0, 0xc000fa20e9, 0x4a, 0x0, 0x0, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:93 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00083c000, 0x37e69a0, 0xc001416180, 0xc0007da578, 0x5400, 0x15000, 0xc0007da610, 0xc000408380, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1182 +0x21e5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000a22000, 0x37e69a0, 0xc001416180, 0xc00083c000, 0x5400, 0x15000, 0xc0007da610, 0xc000408380, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:429 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc000a22000, 0xc000465b00, 0xc000381ac4, 0xc000408380, 0x37e69a0, 0xc001416180, 0xc00083c000, 0x5400, 0x15000, 0xc0007da610, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:331 +0xcf
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:314 +0x1040
```

--- a/pkg/sql/opt/norm/custom_funcs.go
+++ b/pkg/sql/opt/norm/custom_funcs.go
@@ -164,6 +164,14 @@ func (c *CustomFuncs) ArrayType(in memo.RelExpr) types.T {
 	return types.TArray{Typ: inTyp}
 }
 
+// BinaryColType returns the column type of the binary overload for the
+// given operator and operands.
+func (c *CustomFuncs) BinaryColType(op opt.Operator, left, right opt.ScalarExpr) coltypes.T {
+	o, _ := memo.FindBinaryOverload(op, left.DataType(), right.DataType())
+	colType, _ := coltypes.DatumTypeToColumnType(o.ReturnType)
+	return colType
+}
+
 // ----------------------------------------------------------------------
 //
 // Property functions
@@ -1385,6 +1393,29 @@ func (c *CustomFuncs) FoldUnary(op opt.Operator, input opt.ScalarExpr) opt.Scala
 		return nil
 	}
 	return c.f.ConstructConstVal(result, o.ReturnType)
+}
+
+// FoldCast evaluates a cast expression with a constant input. It returns
+// a constant expression as long as the evaluation causes no error.
+func (c *CustomFuncs) FoldCast(input opt.ScalarExpr, colType coltypes.T) opt.ScalarExpr {
+	switch colType.(type) {
+	case *coltypes.TOid:
+		// Save this cast for the execbuilder.
+		return nil
+	}
+
+	datum := memo.ExtractConstDatum(input)
+	texpr, err := tree.NewTypedCastExpr(datum, colType)
+	if err != nil {
+		return nil
+	}
+
+	result, err := texpr.Eval(c.f.evalCtx)
+	if err != nil {
+		return nil
+	}
+
+	return c.f.ConstructConstVal(result, c.ColTypeToDatumType(colType))
 }
 
 // isMonotonicConversion returns true if conversion of a value from FROM to

--- a/pkg/sql/opt/norm/rules/fold_constants.opt
+++ b/pkg/sql/opt/norm/rules/fold_constants.opt
@@ -65,3 +65,19 @@ $result
 )
 =>
 $result
+
+# FoldCast is similar to FoldUnary, but it involves a cast operation. As with
+# FoldUnary, FoldCast applies as long as the evaluation would not cause an
+# error.
+#
+# This rule is marked as low priority so that the FoldNull rules in scalar.opt
+# can run first.
+[FoldCast, Normalize, LowPriority]
+(Cast
+    $input:*
+    $typ:* &
+        (IsConstValueOrTuple $input) &
+        (Succeeded $result:(FoldCast $input $typ))
+)
+=>
+$result

--- a/pkg/sql/opt/norm/rules/numeric.opt
+++ b/pkg/sql/opt/norm/rules/numeric.opt
@@ -3,28 +3,36 @@
 # =============================================================================
 
 # FoldPlusZero folds $left + 0 for numeric types.
+#
+# Note: It is necessary to cast $left to the column type of the binary
+# operation since the type of $left may not match the column type. For example,
+# 1::int + 0::decimal should result in 1::decimal, not 1::int. The execution
+# engine panics when it expects one type but receives another, so this cast is
+# essential. If $left is already of the correct type, the cast will be removed
+# by the EliminateCast rule. Otherwise, if $left is a constant, the cast will
+# be folded away by the FoldCast rule.
 [FoldPlusZero, Normalize]
-(Plus $left:* (Const 0)) => $left
+(Plus $left:* $right:(Const 0)) => (Cast $left (BinaryColType Plus $left $right))
 
 # FoldZeroPlus folds 0 + $right for numeric types.
 [FoldZeroPlus, Normalize]
-(Plus (Const 0) $right:*) => $right
+(Plus $left:(Const 0) $right:*) => (Cast $right (BinaryColType Plus $left $right))
 
 # FoldMinusZero folds $left - 0 for numeric types.
 [FoldMinusZero, Normalize]
-(Minus $left:* (Const 0)) => $left
+(Minus $left:* $right:(Const 0)) => (Cast $left (BinaryColType Minus $left $right))
 
 # FoldMultOne folds $left * 1 for numeric types.
 [FoldMultOne, Normalize]
-(Mult $left:* (Const 1)) => $left
+(Mult $left:* $right:(Const 1)) => (Cast $left (BinaryColType Mult $left $right))
 
 # FoldOneMult folds 1 * $right for numeric types.
 [FoldOneMult, Normalize]
-(Mult (Const 1) $right:*) => $right
+(Mult $left:(Const 1) $right:*) => (Cast $right (BinaryColType Mult $left $right))
 
 # FoldDivOne folds $left / 1 for numeric types.
 [FoldDivOne, Normalize]
-(Div | FloorDiv $left:* (Const 1)) => $left
+(Div | FloorDiv $left:* $right:(Const 1)) => (Cast $left (BinaryColType (OpName) $left $right))
 
 # InvertMinus rewrites -(a - b) to (b - a) if the operand types allow it.
 [InvertMinus, Normalize]

--- a/pkg/sql/opt/norm/testdata/rules/fold_constants
+++ b/pkg/sql/opt/norm/testdata/rules/fold_constants
@@ -572,3 +572,17 @@ values
  ├── key: ()
  ├── fd: ()-->(1)
  └── (false,) [type=tuple{bool}]
+
+# --------------------------------------------------
+# FoldCast
+# --------------------------------------------------
+
+opt expect=FoldCast
+SELECT 1::int/1
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (1,) [type=tuple{decimal}]

--- a/pkg/sql/opt/norm/testdata/rules/numeric
+++ b/pkg/sql/opt/norm/testdata/rules/numeric
@@ -34,6 +34,28 @@ project
       ├── d + d [type=decimal, outer=(4)]
       └── d + d [type=decimal, outer=(4)]
 
+
+# Regression test for #35113.
+opt expect=FoldPlusZero
+SELECT 1::int8 + 0::decimal
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (1,) [type=tuple{decimal}]
+
+opt expect=FoldZeroPlus
+SELECT 0::decimal + 1::int8
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (1,) [type=tuple{decimal}]
+
 # --------------------------------------------------
 # FoldMinusZero
 # --------------------------------------------------
@@ -54,6 +76,17 @@ project
       ├── i + i [type=int, outer=(2)]
       ├── f + f [type=float, outer=(3)]
       └── d + d [type=decimal, outer=(4)]
+
+# Regression test for #35113.
+opt expect=FoldMinusZero
+SELECT 0::int8 - 0::decimal
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (0,) [type=tuple{decimal}]
 
 # --------------------------------------------------
 # FoldMultOne, FoldOneMult
@@ -79,6 +112,27 @@ project
       ├── d + d [type=decimal, outer=(4)]
       └── d + d [type=decimal, outer=(4)]
 
+# Regression test for #35113.
+opt expect=FoldMultOne
+SELECT 2::int8 * 1::decimal
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (2,) [type=tuple{decimal}]
+
+opt expect=FoldOneMult
+SELECT 1::decimal * 2::int8
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (2,) [type=tuple{decimal}]
+
 # --------------------------------------------------
 # FoldDivOne
 # --------------------------------------------------
@@ -95,9 +149,30 @@ project
  ├── scan a
  │    └── columns: i:2(int) f:3(float) d:4(decimal)
  └── projections
-      ├── variable: i [type=int, outer=(2)]
+      ├── i::DECIMAL [type=decimal, outer=(2)]
       ├── variable: f [type=float, outer=(3)]
       └── variable: d [type=decimal, outer=(4)]
+
+# Regression test for #35113.
+opt expect=FoldDivOne
+SELECT 1::int8 / 1::decimal
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (1,) [type=tuple{decimal}]
+
+opt expect=FoldDivOne
+SELECT 1::int8 / 1::int8
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (1,) [type=tuple{decimal}]
 
 # --------------------------------------------------
 # InvertMinus

--- a/pkg/sql/opt/norm/testdata/rules/scalar
+++ b/pkg/sql/opt/norm/testdata/rules/scalar
@@ -1103,14 +1103,10 @@ values
 opt
 SELECT k FROM a WHERE k = ANY '{1,2,3}'::INT[]
 ----
-select
+scan a
  ├── columns: k:1(int!null)
- ├── key: (1)
- ├── scan a
- │    ├── columns: k:1(int!null)
- │    └── key: (1)
- └── filters
-      └── k = ANY '{1,2,3}'::INT8[] [type=bool, outer=(1)]
+ ├── constraint: /1: [/1 - /3]
+ └── key: (1)
 
 # --------------------------------------------------
 # FoldCollate

--- a/pkg/sql/opt/xform/testdata/external/liquibase
+++ b/pkg/sql/opt/xform/testdata/external/liquibase
@@ -359,7 +359,7 @@ project
  │    │    │         │    │    │    │    │    │    │    │    ├── key: (1)
  │    │    │         │    │    │    │    │    │    │    │    └── fd: (1)-->(2,3,5,8,10,13,15,17,20,22,23,26,27), (2,3)-->(1,5,8,10,13,15,17,20,22,23,26,27)
  │    │    │         │    │    │    │    │    │    │    └── filters
- │    │    │         │    │    │    │    │    │    │         └── (c.relkind = 'r'::CHAR) OR (c.relkind = 'f'::CHAR) [type=bool, outer=(17)]
+ │    │    │         │    │    │    │    │    │    │         └── (c.relkind = 'r') OR (c.relkind = 'f') [type=bool, outer=(17)]
  │    │    │         │    │    │    │    │    │    ├── scan n@pg_namespace_nspname_index
  │    │    │         │    │    │    │    │    │    │    ├── columns: n.oid:28(oid!null) n.nspname:29(string!null)
  │    │    │         │    │    │    │    │    │    │    ├── constraint: /29: [/'public' - /'public']

--- a/pkg/sql/opt/xform/testdata/external/navicat
+++ b/pkg/sql/opt/xform/testdata/external/navicat
@@ -362,7 +362,7 @@ sort
       │    │    │         │    │    │    │    │    │    │    │    ├── key: (1)
       │    │    │         │    │    │    │    │    │    │    │    └── fd: (1)-->(2,3,5,8,10,13,15,17,20,22,23,26,27), (2,3)-->(1,5,8,10,13,15,17,20,22,23,26,27)
       │    │    │         │    │    │    │    │    │    │    └── filters
-      │    │    │         │    │    │    │    │    │    │         └── (c.relkind = 'r'::CHAR) OR (c.relkind = 'f'::CHAR) [type=bool, outer=(17)]
+      │    │    │         │    │    │    │    │    │    │         └── (c.relkind = 'r') OR (c.relkind = 'f') [type=bool, outer=(17)]
       │    │    │         │    │    │    │    │    │    ├── scan n@pg_namespace_nspname_index
       │    │    │         │    │    │    │    │    │    │    ├── columns: n.oid:28(oid!null) n.nspname:29(string!null)
       │    │    │         │    │    │    │    │    │    │    ├── constraint: /29: [/'public' - /'public']

